fw/t/unit/test_http_parser.c

```diff
@@ -1321,52 +1284,188 @@ TEST(http_parser, suspicious_x_forwarded_for)
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_CLOSE, req->flags));
+		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_KA, req->flags));
 		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_EXTRA, req->flags));
 	}
+
+	FOR_REQ_SIMPLE("Connection: Close")
+	{
+		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_CLOSE, req->flags));
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_KA, req->flags));
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_EXTRA, req->flags));
+	}
+
+	FOR_REQ_SIMPLE("Connection: close")
+	{
+		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_CLOSE, req->flags));
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_KA, req->flags));
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_EXTRA, req->flags));
+	}
+
+	FOR_REQ_SIMPLE("Connection: Dummy")
+	{
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_CLOSE, req->flags));
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_KA, req->flags));
+		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_EXTRA, req->flags));
+	}
+
+	FOR_REQ_SIMPLE("Connection: Close, Dummy")
+	{
+		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_CLOSE, req->flags));
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_KA, req->flags));
+		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_EXTRA, req->flags));
+	}
+
+	FOR_REQ_SIMPLE("Connection: Keep-Alive, Dummy")
+	{
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_CLOSE, req->flags));
+		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_KA, req->flags));
+		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_EXTRA, req->flags));
+	}
+
+	EXPECT_BLOCK_REQ_SIMPLE("Connection: Close, Keep-Alive");
+
+
+	FOR_REQ_SIMPLE("Connection: closekeep-alive")
+	{
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_CLOSE, req->flags));
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_KA, req->flags));
+		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_EXTRA, req->flags));
+	}
+
+	FOR_REQ_SIMPLE("Connection: keep-alive1")
+	{
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_CLOSE, req->flags));
+		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_KA, req->flags));
+		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_EXTRA, req->flags));
+	}
 }
 
 TEST(http_parser, content_length)
 {
+	char buf[256];
+
+	// Content-Length is mandatory for responses
+	EXPECT_BLOCK_RESP("HTTP/1.1 200 OK\r\n\r\n");
+
+	FOR_REQ("GET / HTTP/1.1\r\n"
+		"Content-Length: 0\r\n"
+		"\r\n")
+	{
+		EXPECT_TRUE(req->content_length == 0);
+	}
+
 	EXPECT_BLOCK_REQ("GET / HTTP/1.1\r\n"
-			  "Content-Length: 0\r\n"
-			  "Content-Length: 0\r\n"
-			  "\r\n");
+			 "Content-Length: 5\r\n"
+			 "\r\n");
+
+	EXPECT_BLOCK_RESP("HTTP/1.0 200 OK\r\n"
+		 	  "Content-Length: 5\r\n"
+		 	  "\r\n");
+
+	FOR_RESP("HTTP/1.0 200 OK\r\n"
+		 "Content-Length: 5\r\n"
+		 "\r\n"
+		 "dummy")
+	{
+		EXPECT_TRUE(resp->content_length == 5);
+	}
+
+	EXPECT_BLOCK_RESP("HTTP/1.0 200 OK\r\n"
+			  "Content-Length: 1000\r\n"
+		 	  "\r\n"
+		 	  "dummy");
+
+	sprintf(buf, "HTTP/1.0 200 OK\r\n"
+		"Content-Length: %lu\r\n"
+		"\r\n"
+		"dummy", ~0UL);
+	EXPECT_BLOCK_RESP(buf);
+
+	// sprintf(buf, "HTTP/1.0 200 OK\r\n"
+	// 	"Content-Length: %lu0\r\n" // +1 digit for overflow
+	// 	"\r\n"
+	// 	"dummy", ~0UL);
+	// EXPECT_BLOCK_RESP(buf);
```

Comment:
Please don't leave such comments in the final version

Reply:
Of course